design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+There are a few options for deploying and hosting the controllers that will perform this work, as explained in the following sections.
+
+#### In-proc Controllers
+
+Multiple controllers could be deployed within a single process.
+For instance, Rook could run one controller per domain of expertise, e.g., ceph-controller, minio-controller, etc.
+Controllers would all watch the same custom resource types via a `SharedInformer` and respond to events via a `WorkQueue` for efficiency and reduced burden on core apiserver.
+Even though all controllers are watching, only the applicable controller responds to and handles an event.
+For example, the user runs `kubectl create cluster.yaml` to create a `cluster.rook.io` instance that has Ceph specific properties.
+All controllers will receive the created event via the `SharedInformer`, but only the Ceph controller will queue and handle it.
+We can consider only loading the controllers the user specifically asks for, perhaps via an environment variable in `rook-operator.yaml`.
+
+Note that this architecture can be used with either API aggregation or CRDs.
+
+##### PROS
+
+* Potentially easier developer burden for new storage backends since there is no need to create a new deployment to host their controller.
+* Less resource burden on K8s cluster since watchers/caches are being shared.
+* Easier migration path to API aggregation in the future, if CRDs usage is continued now.
+
+##### CONS
+
+* All controllers must use same base image since they are all running in the same process.
+* If a controller needs to access a backend specific tool then they will have to schedule a job that invokes the appropriate image.  This is similar to exec’ing a new process but at the cluster level.
+  * We have very limited need for this so far with the Ceph implementation (e.g., creating the initial crush map).
+  * Note this only applies to the controller, not to the backend's daemons.  Those daemons will be running a backend specific image and can directly `exec` to their tools.
+
+#### Operator pods
+
+Each storage backend could have their own operator pod that hosts only that backend's controller, e.g., `ceph-operator.yaml`, `minio-operator.yaml`, etc.
+The user would decide which operators to deploy based on what storage they want to use.
+Each operator pod would watch the same custom resource types with their own individual watchers.
+
+##### PROS
+
+* Each operator can use their own image, meaning they have direct access (through `exec`) to their backend specific tools.
+
+##### CONS
+
+* More difficult migration path to API aggregation in the future.
+* Potentially more resource usage and load on Kubernetes API since watchers will not be shared.
+* Potentially more developer burden as all backends have to write their own deployment/host to manage their individual pod.
```

Comment:
The developer burden of this seems much lower to me than the burden of factoring out the backend-specific commands into jobs.

Reply:
I need to look how rook currently handles HA. Classic HA services would need very close integration with backend functionality. That is, iSCSI/HA, NFS/HA or SMB/HA. For some users fail-over time under < 10 seconds is a hard requirement. Within given time backend would need to move its own objects from stand by to active, it has to be fast..